File: typescript/factory_method/transportation/transport_unit.ts
export interface Transport_unit {
  deliver(): string
}

export class Truck implements Transport_unit {
  public deliver() {
    console.log('truck')
    return 'truck'
  }
}

export class Ship implements Transport_unit {
  public deliver(): string {
    console.log('ship')
    return 'ship'
  }
}

export class Plane implements Transport_unit {
  public deliver(): string {
    console.log('plane')
    return 'plane'
  }
}

File: dart/factory_method/transportation/bin/transport_unit.dart
abstract class TransportUnit {
  String deliver();
}

class Truck implements TransportUnit {
  @override
  String deliver() {
    print('truck');
    return 'truck';
  }
}

class Ship implements TransportUnit {
  @override
  String deliver() {
    print('ship');
    return 'ship';
  }
}

class Plane implements TransportUnit {
  @override
  String deliver() {
    print('plane');
    return 'plane';
  }
}
